x/auth/middleware/persist.go

```diff
@@ -0,0 +1,76 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+)
+
+type persistHandler struct {
```

Comment:
I spent some time thinking of a test, LMK if the below works:

```go
package middleware_test

import (
	"context"
	"fmt"
	"math"

	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
	"github.com/cosmos/cosmos-sdk/testutil/testdata"
	sdk "github.com/cosmos/cosmos-sdk/types"
	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
	"github.com/cosmos/cosmos-sdk/types/tx"
	"github.com/cosmos/cosmos-sdk/x/auth/middleware"
)

func (s *MWTestSuite) TestPersist() {
	testcases := []struct {
		name         string
		gasToConsume uint64 // gas to consume in the msg execution
		expErr       bool
	}{
		{"more than block gas meter", blockMaxGas, true},
		{"less than block gas meter", 10, false},
	}

	for _, tc := range testcases {
		tc := tc
		s.Run(tc.name, func() {
			ctx := s.SetupTest(true) // setup
			txBuilder := s.clientCtx.TxConfig.NewTxBuilder()

			// testMsgTxHandler is a test txHandler that handles one single TestMsg,
			// consumes the given `tc.gasToConsume`, and sets the bank store "ok" key to "ok".
			var testMsgTxHandler = customTxHandler{func(ctx context.Context, req tx.Request) (tx.Response, error) {
				msg, ok := req.Tx.GetMsgs()[0].(*testdata.TestMsg)
				if !ok {
					return tx.Response{}, fmt.Errorf("expected %T, got %T", (*testdata.TestMsg)(nil), msg)
				}

				sdkCtx := sdk.UnwrapSDKContext(ctx)
				sdkCtx.KVStore(s.app.GetKey("bank")).Set([]byte("ok"), []byte("ok"))
				sdkCtx.GasMeter().ConsumeGas(tc.gasToConsume, "TestMsg")
				return tx.Response{}, nil
			}}

			txHandler := middleware.ComposeMiddlewares(
				testMsgTxHandler,
				middleware.NewTxDecoderMiddleware(s.clientCtx.TxConfig.TxDecoder()),
				middleware.GasTxMiddleware,
				middleware.PersistTxMiddleware,
				middleware.RecoveryTxMiddleware,
			)

			// msg and signatures
			priv1, _, addr1 := testdata.KeyTestPubAddr()
			msg := testdata.NewTestMsg(addr1)
			atomCoin := sdk.NewCoin("atom", sdk.NewInt(150))
			feeAmount := sdk.NewCoins(atomCoin)
			var gasLimit uint64 = math.MaxUint64 // no limit on sdk.GasMeter
			s.Require().NoError(txBuilder.SetMsgs(msg))
			txBuilder.SetFeeAmount(feeAmount)
			txBuilder.SetGasLimit(gasLimit)

			privs, accNums, accSeqs := []cryptotypes.PrivKey{priv1}, []uint64{0}, []uint64{0}
			testTx, _, err := s.createTestTx(txBuilder, privs, accNums, accSeqs, ctx.ChainID())
			s.Require().NoError(err)

			_, err = txHandler.DeliverTx(sdk.WrapSDKContext(ctx), tx.Request{Tx: testTx})

			bankStore := ctx.KVStore(s.app.GetKey("bank"))
			okValue := bankStore.Get([]byte("ok"))

			if tc.expErr {
				s.Require().Error(err)
				s.Require().True(sdkerrors.IsOf(err, sdkerrors.ErrOutOfGas))
				s.Require().Empty(okValue)

			} else {
				s.Require().NoError(err)
				s.Require().Equal([]byte("ok"), okValue)
			}
		})
	}

}
```

Reply:
thanks, I've adapted your test code, adding checks on fee deduction.